router/internal/eventorder/eventorder.go

```diff
@@ -0,0 +1,255 @@
+func (b *Barrier) Enter(userID string, jobID int64) (accepted bool, previousFailedJobID *int64) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	barrier, ok := b.barriers[userID]
+	if !ok {
+		return true, nil // no barrier, accept
+	}
+
+	// if there is a failed job in the barrier, only this job can enter the barrier
+	if barrier.failedJobID != nil {
+		failedJob := *barrier.failedJobID
+		return jobID == failedJob, &failedJob
+	}
+
+	// if the concurrency limiter is enabled, only the configured number of concurrent jobs can enter the barrier
+	if barrier.concurrencyLimiter != nil {
+		barrier.mu.Lock()
+		defer barrier.mu.Unlock()
+		if _, ok := barrier.concurrencyLimiter[jobID]; ok {
+			return true, nil // if the job is already in the concurrent jobs map, accept it (poor job... it forgot to notify the barrier before leaving!)
+		}
+		if len(barrier.concurrencyLimiter) >= b.concurrencyLimit {
+			return false, nil // if the concurrent jobs map is full, reject it
+		}
+
+		// add the job to the concurrent jobs map and accept it
+		barrier.concurrencyLimiter[jobID] = struct{}{}
+		return true, nil
+	}
+
+	return true, nil
+}
+
+// Wait returns true if the job for this userID shouldn't continue, but wait (transition to a waiting state)
+func (b *Barrier) Wait(userID string, jobID int64) (wait bool, previousFailedJobID *int64) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	barrier, ok := b.barriers[userID]
+	if !ok {
+		return false, nil // no barrier, don't wait
+	}
+	if barrier.failedJobID != nil {
+		failedJob := *barrier.failedJobID
+		if failedJob > jobID {
+			panic(fmt.Errorf("previousFailedJob:%d > jobID:%d", failedJob, jobID))
+		}
+		return jobID > failedJob, &failedJob // wait if this is not the failed job
+	}
+	// no faild job, don't wait
+	return false, nil
+}
+
+// StateChanged must be called at the end, after the job state change has been persisted.
+// The only exception to this rule is when a job has failed in a retryable manner, in this scenario you should notify the barrier immediately after the failure.
+// An [ErrUnsupportedState] error will be returned if the state is not supported.
+func (s *Barrier) StateChanged(userID string, jobID int64, state State) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var command command
+	jsCmd := &cmd{userID: userID, jobID: jobID}
+	switch state {
+	case SucceededState:
+		command = &jobSucceededCmd{jsCmd}
+	case FailedState:
+		command = &jobFailedCommand{jsCmd}
+	case AbortedState:
+		command = &JobAbortedCommand{jsCmd}
+	case WaitingState:
+		command = jsCmd
+	default:
+		return ErrUnsupportedState
+	}
+
+	if command.enqueue(s.barriers) {
+		s.queue = append(s.queue, command)
+	} else {
+		command.execute(s.barriers)
+	}
+	return nil
+}
+
+// Sync applies any enqueued commands to the barrier's state. It should be called at the beginning of every new iteration of the main loop
+func (s *Barrier) Sync() int {
```

Comment:
Any reason why we are using two receiver names for the same struct?

Reply:
nope, thanks for spotting it 👍